src/Controller/Traits/Code2fTrait.php

```diff
@@ -0,0 +1,199 @@
+        ]);
+    }
+
+    /**
+     * Show Code2f register start step
+     *
+     * @return \Cake\Http\Response|null
+     */
+    public function code2fRegister()
+    {
+        $data = $this->getCode2fData();
+        if (!$data['valid']) {
+            return $this->redirectWithQuery([
+                'action' => 'login',
+            ]);
+        }
+        $field = Configure::read('Code2f.type');
+        $this->set('field', $field);
+        if ($this->getRequest()->is(['post', 'put'])) {
+
+            $value = $this->getRequest()->getData($field);
+            if ($data['field'] === Code2fAuthenticationCheckerInterface::CODE2F_TYPE_PHONE && !preg_match('/^\+[1-9]\d{1,14}$/i', $value)) {
+                $this->Flash->error(__d('cake_d_c/users', 'Invalid phone number: Format must be +1234567890'));
+            } else {
+                $data['user'][$field] = $value;
+                $user = $this->getUsersTable()->saveOrFail($data['user'], ['checkRules' => false]);
+                $this->getRequest()->getSession()->write(AuthenticationService::CODE2F_SESSION_KEY, $user);
+                $data['registration'] = true;
+            }
+        }
+        if ($data['registration']) {
+            return $this->redirectWithQuery([
+                'action' => 'code2fAuthenticate',
+            ]);
+        }
+        $this->viewBuilder()->setLayout('CakeDC/Users.login');
+    }
+
+    /**
+     * Show code2f authenticate start step
+     *
+     * @return \Cake\Http\Response|null
+     */
+    public function code2fAuthenticate()
+    {
+        $data = $this->getCode2fData();
+        if (!$data['valid']) {
+            return $this->redirectWithQuery(Configure::read('Auth.AuthenticationComponent.loginAction'));
+        }
+        if (!$data['registration']) {
+            return $this->redirectWithQuery([
+                'action' => 'code2fRegister',
+            ]);
+        }
+        /** @var OtpCodesTable $OtpCodes */
+        $OtpCodes = TableRegistry::getTableLocator()->get('CakeDC/Users.OtpCodes');
+        $resend = $this->getRequest()->is(['post', 'put']) && $this->getRequest()->getQuery('resend');
+        if ($this->getRequest()->is(['post', 'put']) && !$resend) {
+            try {
+                $result = $OtpCodes->validateCode2f($data['user']['id'], $this->getRequest()->getData('code'));
+                if (!$result) {
+                    $this->Flash->error(__d('cake_d_c/users', 'The code entered is not valid, please try again or resend code.'));
+                }
+                    $this->request->getSession()->delete(AuthenticationService::CODE2F_SESSION_KEY);
+                    $this->request->getSession()->write(TwoFactorAuthenticator::USER_SESSION_KEY, $data['user']);
+                    return $this->redirectWithQuery(Configure::read('Auth.AuthenticationComponent.loginAction'));
+            } catch (\Exception $e) {
+                $this->Flash->error($e->getMessage());
```

Comment:
Are we sure that we will not show bad error messages like database error or expose file path?

Reply:
only specific exceptions are catched now